Preliminary ROCC support

jerryz123 opened the pull request:
Currently this supports a subset of the RoCC interface. Should be enough for Hwacha.

ccelio commented on src/main/scala/exu/core.scala:
```diff
@@ -1477,44 +1504,23 @@ class BoomCore(implicit p: Parameters, edge: freechips.rocketchip.tilelink.TLEdg
-   io.fpu <> DontCare
-   io.fpu.valid := false.B
-   io.fpu.inst := 0.U
+   if (usingRoCC)
```
What happens if !usingRoCC? Do you still tie off the io.rocc signals?

jerryz123 replied:
No, but is it necessary to? They don't drive anything in that case.

ccelio replied:
I don't think it's good practice to leave I/Os dangling---and I've occasionally be surprised when I thought IOs would get trimmed and they didn't (fields within the MicroOp has bitten me here).

Related, I _strongly_ recommend that all modules, including BoomCore, add `override val compileOptions = chisel3.core.ExplicitCompileOptions.NotStrict.copy(explicitInvalidate = true)` to catch undriven I/Os. That will force you to specify DontCare if you want to leave stuff undriven. Importing `chisel3` isn't strong enough here.

jerryz123 commented:
This seems to work with Hwacha.

jerryz123 commented:
@bkorpan I reverted the change to disambiguate rob ordering by head, instead of tail. Instead I fixed the bug where `rob_tail_idx` is inaccurate. Now `rob_tail_idx` has entry granularity, rather than row granularity.

jerryz123 commented on src/main/scala/exu/rob.scala:
```diff
@@ -799,26 +818,36 @@ class Rob(
+   }
+   .elsewhen (io.enq_valids.asUInt =/= 0.U && io.enq_partial_stall)
+   {
+      rob_tail_lsb := PriorityEncoder(~io.enq_valids.asUInt)
```
@ccelio @bkorpan Any thoughts on how the tail bank idx bits are calculated here?

abejgonzalez commented on docs/sections/reorder-buffer.rst:
```diff
@@ -193,3 +193,17 @@ Load/Store Unit, and are treated as exceptions in the BOOM pipeline
+Point of No Return (PNR)
+------------------------
+
+The point-of-no-return head runs ahead of the ROB commit head, marking the
```
"marking the next instruction which might not commit and cause misspeculation"

abejgonzalez replied:
Sentence sounds wierd? "not commit and cause misspeculation"

abejgonzalez commented on src/main/scala/common/rocc.scala:
```diff
@@ -0,0 +1,69 @@
+package boom.common
```
Can we add the proper heading here.

abejgonzalez commented on docs/sections/execution-stages.rst:
```diff
@@ -322,3 +322,35 @@ an exception as part of executing a CSR instruction (e.g., a syscall).
 
 .. [7]
    The CSRFile is a Rocket component.
+
```
Is it possible to get a block diagram of the connection to the ROB, queues, etc? (Not going to block the PR on this though)

jerryz123 replied:
I added the RoCC thing to the basic block diagram we have on lucidchart

abejgonzalez replied:
Update in the docs then?

abejgonzalez commented on src/main/scala/common/rocc.scala:
```diff
@@ -0,0 +1,69 @@
+
+import boom.lsu._
+
+/*
```
Are we adhering to the following syntax for multi-line comments

```
/**
 * This is a comment
 */
```

abejgonzalez commented on src/main/scala/exu/fp-pipeline.scala:
```diff
@@ -168,7 +168,6 @@ class FpPipeline(implicit p: Parameters) extends BoomModule()(p) with tile.HasFP
    //-------------------------------------------------------------
 
    exe_units.map(_.io.brinfo := io.brinfo)
-   exe_units.map(_.io.com_exception := io.flush_pipeline)
```
Why remove this?

jerryz123 replied:
I modified the execution unit IO such that only execution units which require the com_exception signal will have it in the IO

In general I'm trying to move towards IO bundles without extraneous signals.

abejgonzalez replied:
I see. I think this makes sense, although, Im not fully on board with using `null` to create specific bundles. For now, I think this is fine. The alternative is to use the `Some/None/etc` construct that Scala provides.

abejgonzalez commented on src/main/scala/exu/rob.scala:
```diff
@@ -230,25 +232,27 @@ class Rob(
-   val empty = Wire(Bool())
+   val rob_pnr      = RegInit(0.U(log2Ceil(NUM_ROB_ROWS).W))
+   val rob_pnr_lsb  = RegInit(0.U((1 max log2Ceil(width)).W))
+   val rob_pnr_idx  = if (width == 1) rob_pnr  else Cat(rob_pnr , rob_pnr_lsb)
```
Nitpick: Spacing doesn't match rest of coding style

jerryz123 replied:
The spacing is to match `rob_head_idx` and `rob_tail_idx`.

abejgonzalez replied:
But those vals are a couple lines away. I think the better style is to match the things below and above. So in this case, there is no similar val above or below, so you revert to the normal spacing.

abejgonzalez commented on src/main/scala/util/util.scala:
```diff
@@ -353,6 +342,19 @@ object AgePriorityEncoder
   }
 }
 
+// is i0 older than i1? (closest to zero). Provide the tail_ptr to the
```
Put the comment inside the `/** */` comment

abejgonzalez commented on src/main/scala/exu/execution-units/rocc.scala:
```diff
@@ -0,0 +1,209 @@
+//******************************************************************************
+// Copyright (c) 2013 - 2018, The Regents of the University of California (Regents).
```
2019

abejgonzalez commented on src/main/scala/exu/execution-units/rocc.scala:
```diff
@@ -0,0 +1,209 @@
+
+import freechips.rocketchip.config.Parameters
+import freechips.rocketchip.tile.{RoCCCoreIO, RoCCInstruction}
+import freechips.rocketchip.rocket._
```
Space between rocket and boom imports

abejgonzalez commented on src/main/scala/exu/execution-units/rocc.scala:
```diff
@@ -0,0 +1,209 @@
+
+/**
+ * IO bundle representing the different signals to interact with the RoCC
+ *
```
Extra space
